# PEtracer 2025

This repository contains the code to reproduce all analyses and figures from the manuscript 
"High-resolution spatial mapping of cell state and lineage dynamics in vivo with PETracer".

![](https://github.com/jweissmanlab/PETracer_Paper/blob/main/img.png)
# Setup

# Data availability

* Processed data is available on [Figshare](https://figshare.com/s/8e9d573deca3d44235fe)
* Single-cell RNA-seq data is available on [GEO](Add link here)
* All other sequencing data is available on [SRA](Add link here)

# Image processing

# Colony lineage tracing

The [colony_tracing](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing) directory 
contains code for processing and analyzing single-cell lineage tracing from colonies generated by 
sparsely seeding 4T1 cells onto a coverslip.

### Data processing

After processing raw images as described in the "Image processing" section the **colony_process_lineage.ipynb** 
notebook was used to segment colonies, perform quality control, and reconstruct phylogenies.

### Plotting

All tumor plots can be generated by running
```bash
python colony_tracing/plot.py
```
after the following files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) are downloaded 
and placed in the [colony_tracing/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing/data) directory:

* colony_tracing.h5td
* colony_polygons.json

# 4T1 tumor lineage tracing

The [tumor_tracing](https://github.com/jweissmanlab/PETracer_Paper/tree/main/tumor_tracing) directory 
contains code for processing and analyzing single-cell transcriptomic and lineage tracing data 
from the 4T1 syngeneic mouse model of tumor metastasis.

### Data processing

After processing raw images as described in the "Image processing" section, the following notebooks were used to generate the mouse 1 data:

1. **M1_resolVI_training.ipynb** - trains resolVI model to classify cell types and filter out doublets.
2. **M1_process_MERFISH.ipynb** - performs quality control and annotation of the MERFISH data.
3. **M1_segment_tumors.ipynb** - aligns tumor sections, segments tumors, and calculate spatial statistics.
4. **M1_process_lineage.ipynb** - performs quality control of lineage data and reconstructs phylogenies.

The same process was repeated for the mouse 2 and 3 data, except a new resolVI model was not trained for mouse 2 since the library is shared with mouse 1.

### Plotting

All tumor plots can be generated by running
```bash
python tumor_tracing/plot.py
```
after the following files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) are downloaded 
and placed in the [tumor_tracing/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/tumor_tracing/data) directory:

* 10x_4T1_primary.h5ad
* M1_tumor_tracing.h5td
* M1_polygons_grid.json
* M2_tumor_tracing.h5td
* M2_polygons.json
* M3_tumor_tracing.h5td
* M3_polygons_grid.json