# PEtracer 2025

This repository contains the code to reproduce all analyses and figures from the manuscript 
"High-resolution spatial mapping of cell state and lineage dynamics in vivo with PETracer".

![](https://github.com/jweissmanlab/PETracer_Paper/blob/main/img.png)
# Setup

# Data availability

* Processed data is available on [Figshare](https://figshare.com/s/8e9d573deca3d44235fe)
* Single-cell RNA-seq data is available on [GEO](Add GEO link)
* All other sequencing data is available on [SRA](Add SRAN link)

# Simulation

The [simulation](https://github.com/jweissmanlab/PETracer_Paper/tree/main/simulation) directory 
contains code for simulating lineage tracing data with a variety of parameters. To run simulations:

```bash 
python simulation/simulate.py
```

To generate plots:
```bash
python simulation/plot.py
```

# pegRNA variant kinetics

The [kinetics](https://github.com/jweissmanlab/PETracer_Paper/tree/main/kinetics) directory 
contains code for processing and analyzing 10x data for 4T1 and B16 cells transduced a library
of pegRNA variants to test editing kinetics.

### Data processing

10x data was processed on a Linux HPC cluster with SLURM, Python 3.11, and Cellranger 7.1.0 installed.
Processed files can be generated with the following steps:

1. Run Cellranger and call alleles using bam files.
```bash 
sbatch kinetics/cellranger.slurm
sbatch kinetics/call_alleles.slurm
```
2. **process_4T1_10x.ipynb** - perform quality control, call pegRNA variants, and determine edit fraction for 4T1 cells.
3. **process_B16F10_10x.ipynb** - perform quality control, call pegRNA variants, and determine edit fraction for B16F10 cells.


after downloading the 10x fastq files listed in [manifest.txt](https://github.com/jweissmanlab/PETracer_Paper/tree/main/kinetics/fastq/manifest.txt) from [GEO](Add GEO link) and placing them in the [kinetics/fastq](https://github.com/jweissmanlab/PETracer_Paper/tree/main/preedited/fastq) directory.

### Analysis

All kinetics analysis and plots can be generated by running
```bash
python kinetics/estimate_rate.py
python kinetics/plot.py
```
after processing the raw data or downloading the processed files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) and placing them in [kinetics/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/kinetics/data) directory:

* 4T1_kinetics_alleles.csv
* 4T1_kinetics_cells.csv
* 4T1_kinetics.h5ad
* B16F10_kinetics_alleles.csv
* B16F10_kinetics_cells.csv
* B16F10_kinetics.h5ad

# Image processing

# Predifined lineage mark validation

The [preedited](https://github.com/jweissmanlab/PETracer_Paper/tree/main/preedited) directory 
contains code for processing and analyzing 10x and imaging-based readout of lineage tracing data
from cells with predefined linkage between intBCs and lineage marks.

### Data processing

Imaging data was processed as described in the "Image processing" section. 10x data was processed on 
a Linux HPC cluster with SLURM, Python 3.11, and Cellranger 7.1.0 installed.
Processed files can be generated with the following steps:

1. For 10x run Cellranger and call alleles using bam files.
```bash 
sbatch preedited/cellranger.slurm
sbatch preedited/call_alleles.slurm
```
2. **process_10x_invitro.ipynb** - perform quality control for 10x in vitro data.
3. **process_merfish_invitro.ipynb** - perform quality control for imaging in vitro data.
4. **process_merfish_zombie.ipynb** - perform quality control for imaging in vitro data using the zombie protocol.
5. **process_merfish_invivio.ipynb** - perform quality control for imaging in vivo data.

after downloading the 10x fastq files listed in [manifest.txt](https://github.com/jweissmanlab/PETracer_Paper/tree/main/preedited/fastq/manifest.txt) from [GEO](Add GEO link) and placing them in the [preedited/fastq](https://github.com/jweissmanlab/PETracer_Paper/tree/main/preedited/fastq) directory.

### Analysis

All preedited analysis and plots can be generated by running
```bash
python preedited/plot.py
```
after processing the raw data or downloading the processed files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) and placing them in [peedited/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing/data) directory:

* preedited_10x_invitro_alleles.csv
* preedited_10x_invitro.h5ad
* preedited_merfish_invitro_alleles.csv
* preedited_merfish_invitro_cells.json
* preedited_merfish_invivo_alleles.csv
* preedited_merfish_invivo_cells.json
* preedited_merfish_zombie_alleles.csv
* preedited_merfish_zombie_cells.json

# Barcoded lineage tracing

The [barcoded_tracing](https://github.com/jweissmanlab/PETracer_Paper/tree/main/barcoded_tracing) directory 
contains code for processing and analyzing 10x single-cell lineage tracing for clones with puro and blast-linked
static barcodes serving as independent validation of phylogenetic relationships.

### Data processing

Data processing was performed on a Linux HPC cluster with SLURM, Python 3.11, and Cellranger 7.1.0 installed.
Processed files can be generated with the following steps:

1. Run Cellranger and call alleles using bam files.
```bash 
sbatch barcoded_tracing/cellranger.slurm
sbatch barcoded_tracing/call_alleles.slurm
```
2. **process_10x.ipynb** - performs quality control, phylogenetic reconstruction, and processing of barcode data.

after downloading the files listed in [manifest.txt](https://github.com/jweissmanlab/PETracer_Paper/tree/main/barcoded_tracing/fastq/manifest.txt) from [GEO](Add GEO link) and placing them in the [barcoded_tracing/fastq](https://github.com/jweissmanlab/PETracer_Paper/tree/main/barcoded_tracing/fastq) directory.

### Analysis

All barcoded tracing analysis and plots can be generated by running
```bash
python barcoded_tracing/evaluate.py
python barcoded_tracing/plot.py
```
after processing the raw data or downloading the processed files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) and placing them in [colony_tracing/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing/data) directory:

* barcoded_tracing_clone_1.h5td
* barcoded_tracing_clone_2.h5td
* barcoded_tracing_clone_3.h5td
* barcoded_tracing_clone_4.h5td
* barcoded_tracing_clone_5.h5td
* barcoded_tracing_clone_6.h5td
* barcoded_tracing_alleles.csv

# Colony lineage tracing

The [colony_tracing](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing) directory 
contains code for processing and analyzing single-cell lineage tracing from colonies generated by 
sparsely seeding 4T1 cells onto a coverslip.

### Data processing

After processing raw images as described in the "Image processing" section the **colony_process_lineage.ipynb** 
notebook was used to segment colonies, perform quality control, and reconstruct phylogenies.

### Analysis

All colony plots can be generated by running
```bash
python colony_tracing/plot.py
```
after the following files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) are downloaded 
and placed in the [colony_tracing/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/colony_tracing/data) directory:

* colony_tracing.h5td
* colony_polygons.json

# 4T1 tumor lineage tracing

The [tumor_tracing](https://github.com/jweissmanlab/PETracer_Paper/tree/main/tumor_tracing) directory 
contains code for processing and analyzing single-cell transcriptomic and lineage tracing data 
from the 4T1 syngeneic mouse model of tumor metastasis.

### Data processing

After processing raw images as described in the "Image processing" section, the following notebooks were used to generate the mouse 1 data:

1. **M1_resolVI_training.ipynb** - trains resolVI model to classify cell types and filter out doublets.
2. **M1_process_MERFISH.ipynb** - performs quality control and annotation of the MERFISH data.
3. **M1_segment_tumors.ipynb** - aligns tumor sections, segments tumors, and calculate spatial statistics.
4. **M1_process_lineage.ipynb** - performs quality control of lineage data and reconstructs phylogenies.

The same process was repeated for the mouse 2 and 3 data, except a new resolVI model was not trained for mouse 2 since the library is shared with mouse 1.

### Analysis

All tumor plots can be generated by running
```bash
python tumor_tracing/plot.py
```
after the following files from [Figshare](https://figshare.com/s/8e9d573deca3d44235fe) are downloaded 
and placed in the [tumor_tracing/data](https://github.com/jweissmanlab/PETracer_Paper/tree/main/tumor_tracing/data) directory:

* 10x_4T1_primary.h5ad
* M1_tumor_tracing.h5td
* M1_polygons_grid.json
* M2_tumor_tracing.h5td
* M2_polygons.json
* M3_tumor_tracing.h5td
* M3_polygons_grid.json